src/compiler/core.ts

```diff
@@ -1492,6 +1492,46 @@ namespace ts {
         return createMultiMap() as UnderscoreEscapedMultiMap<T>;
     }
 
+    export function createQueue<T>(items?: readonly T[]): Queue<T> {
+        const elements: (T | undefined)[] = items?.slice() || [];
+        let headIndex = 0;
+
+        function isEmpty() {
+            return headIndex === elements.length;
+        }
+
+        function enqueue(...items: T[]) {
+            elements.push(...items);
+        }
+
+        function dequeue(): T {
+            if (isEmpty()) {
+                throw new Error("Queue is empty");
+            }
+
+            const result = elements[headIndex] as T;
+            elements[headIndex] = undefined; // Don't keep referencing dequeued item
+            headIndex++;
+
+            // If more than half of the queue is empty, copy the remaining elements to the
+            // front and shrink the array (unless we'd be saving fewer than 100 slots)

```

Comment:
out of curiosity - if you only need a simple interface of those 3 methods, wouldn't it be somewhat more clear to go for a simple linked list than keeping the internal representation as array, track the `head` pointer and perform this resizing dance?

does using an array here have an advantage over a linked list of objects like this?
```ts
interface Item<T> {
  value: T;
  next: Item<T> | null;
}
```

Reply:
I did not specifically measure, but my expectation would be that:
1. A packed array has better memory locality
2. A linked list would generate a lot more short-lived allocations
3. A single array access is faster than two member access (.next.value)

Reply:
Looks similar in my particular repro, but I still suspect you'd be bitten by GC in long-running scenarios.  Also, this approach is a smaller change from the array/shift code that I'm replacing.

Reply:
Thanks for the answer!

> A linked list would generate a lot more short-lived allocations

This one I don't understand - don't both solutions allocate the very same number of objects? The only difference I see is that you always hold the array in memory, even if it's empty but that shouldn't matter as it's just acting as "proxy" for the actual list. In both solutions, we have some kind of an instance of a class (ArrayBasedQueue vs LinkedList) and N enqueued items (where `arrayBasedQueue` holds an additional array but that's constant and doesn't matter).

Reply:
Sorry, I wasn't very clear - I was referring to the data structure overhead and not the collection members.  With the array, there's a closure object holding index and the array, plus the array itself.  With the linked list, there's a closure object holding the head and tail pointers (ish - there are other implementations) and there's an instance of `Item` wrapping each queue element.

Reply:
It seems I had a brain fart and somehow skipped over the fact that LinkedList wraps each item in an additional container~ object so it has 2N allocations whereas your ArrayBasedQueue has N. Thanks for bearing with me!

Reply:
No worries - always good to question assumptions.